src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -57,51 +60,150 @@ public override Task NewSolutionSnapshotAsync(Solution solution, CancellationTok
         {
-            return SpecializedTasks.EmptyImmutableArray<DiagnosticData>();
+            var diagnostics = await GetDiagnosticsAsync(solution, projectId, documentId, cancellationToken).ConfigureAwait(false);
+            return diagnostics.Where(d => diagnosticIds.Contains(d.Id)).ToImmutableArrayOrEmpty();
         }
 
-        public override Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsForIdsAsync(Solution solution, ProjectId projectId = null, ImmutableHashSet<string> diagnosticIds = null, CancellationToken cancellationToken = default(CancellationToken))
+        public override async Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsForIdsAsync(Solution solution, ProjectId projectId = null, ImmutableHashSet<string> diagnosticIds = null, CancellationToken cancellationToken = default(CancellationToken))
         {
-            return SpecializedTasks.EmptyImmutableArray<DiagnosticData>();
+            var diagnostics = await GetDiagnosticsForIdsAsync(solution, projectId, null, diagnosticIds, cancellationToken).ConfigureAwait(false);
+            return diagnostics.Where(d => d.DocumentId == null).ToImmutableArray();
         }
 
-        public override Task<bool> TryAppendDiagnosticsForSpanAsync(Document document, TextSpan range, List<DiagnosticData> diagnostics, CancellationToken cancellationToken)
+        public override async Task<bool> TryAppendDiagnosticsForSpanAsync(Document document, TextSpan range, List<DiagnosticData> result, CancellationToken cancellationToken)
         {
-            return SpecializedTasks.False;
+            result.AddRange(await GetDiagnosticsForSpanAsync(document, range, cancellationToken).ConfigureAwait(false));
+            return true;
         }
 
-        public override Task<IEnumerable<DiagnosticData>> GetDiagnosticsForSpanAsync(Document document, TextSpan range, CancellationToken cancellationToken)
+        public override async Task<IEnumerable<DiagnosticData>> GetDiagnosticsForSpanAsync(Document document, TextSpan range, CancellationToken cancellationToken)
         {
-            return SpecializedTasks.EmptyEnumerable<DiagnosticData>();
+            var diagnostics = await GetDiagnosticsAsync(document.Project.Solution, document.Project.Id, document.Id, cancellationToken).ConfigureAwait(false);
+            return diagnostics.Where(d => range.IntersectsWith(d.TextSpan));
+        }
+
+        private async Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsAsync(Project project, CancellationToken cancellationToken)
+        {
+            if (project == null)
+            {
+                return ImmutableArray<DiagnosticData>.Empty;
+            }
+
+            var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+            var analyzers = _analyzerManager.CreateDiagnosticAnalyzers(project);
```

Comment:
Should this be instead renamed to "GetDiagnosticAnalyzers" as it isn't creating them for each invocation?